### DISCUSSION TITLE
Hinge Constraint for door swing

### DISCUSSION THREAD

**Discussion by wending10** [jrouwe/JoltPhysics#950] [Q&A] [answered]
I have implemented a simple hingle constraint with reference from the sample provided in PoweredHingeConstraintTest.cpp

https://github.com/jrouwe/JoltPhysics/assets/90673673/d8cb838a-b9fb-4046-a024-2f60a0862054

I would like to in scripts (for example upon keypress, it then trigger the door rotation), it then start rotating about the given point. I have explored ```SetRotation``` ```SetAngularVelocity``` but to no avail. it simply does not want to rotate. 

TLDR: I want to rotate door about a given point, set the hinge constraint, however I do not know how to rotate through scripts

**Comment by jrouwe** [comment]
Have you tried using motors?

You can set the velocity:

```
mConstraint->SetMotorState(EMotorState::Velocity);
mConstraint->SetTargetAngularVelocity(DegreesToRadians(25));
```

alternatively you can directly specify the target rotation using:

```
mConstraint->SetMotorState(EMotorState::Position);
mConstraint->SetTargetAngle((DegreesToRadians(somevalue));
```

B.t.w. SetAngularVelocity should also work, maybe you're not setting a velocity in the right direction (the blue axis would be the z-axis so an angular velocity of Vec3(0, 0, something).

**Reply by wending10** [reply]
Thanks for the prompt reply! My problem is more on that it is only able to rotate due to setting either the velocity or the target rotation in the initialise function. The video that i attached did have the code:
```
HingeConstraint* hinge = static_cast<HingeConstraint*>(hingeSettings.Create(*boxBodyHinge, *boxBody));
hinge->SetMotorState(EMotorState::Velocity);
hinge->SetTargetAngularVelocity(15.f * Mathf::Deg2Rad);
m_pSystem->AddConstraint(hinge);
```
hence it rotated as shown in the sample

my question is more on that what if I want it to rotate on a trigger event? with reference like a door, say if I press "E to interact", it should thne swing outwards 90 degree

how do i:
1) Access the particular body constraint to set the target angle/velocity
2) if not, use functions such as ```setRotation``` or ```setAngularVelocity```
-> 2.1) i tried using setAngularVelocity, but it seems that it want to rotate on its center position and not the pivot point, how do I get it to rotate around the hinge point?
-> 2.2) unless i am suppose to rotate the hinge? and not the door?

**Reply by jrouwe** [reply, marked as answer]
> Access the particular body constraint to set the target angle/velocity

You need to keep a pointer to this constraint in your own code or you can iterate over the return value of ```PhysicsSystem::GetConstraints()``` to find the right one (e.g. based on the contents of ```Constraint::GetUserData()```.

> i tried using setAngularVelocity, but it seems that it want to rotate on its center position and not the pivot point, how do I get it to rotate around the hinge point?

You would need to set the center of mass of the object to match where you want it to rotate about, to do this you wrap your ```Shape``` in an ```OffsetCenterOfMassShape```. But I would personally not work with SetAngularVelocity but use motors as mentioned before or you could just apply a linear impulse (```Body::AddImpulse```) to the door to 'push' it open.

> unless i am suppose to rotate the hinge? and not the door?

No you should not rotate the hinge.

**Reply by wending10** [reply]
Thank you for the input, ```Constraint::GetUserData``` solve my issue, after i SetUserData in HingeSetting, im able to access it via the iteration that you mention

```
JPH::Constraints constraintList = m_pSystem->GetConstraints();
for (auto constraint : constraintList)
{
	if (constraint->GetUserData() == _rigidbody->GetBodyID())
	{
		JPH::HingeConstraint* hinge = static_cast<JPH::HingeConstraint*>(constraint.GetPtr());
		hinge->SetTargetAngle((15.f) * Mathf::Deg2Rad);
	}

}
```

**Comment by jrouwe** [comment]
Another question: Are you sure the door didn't go to sleep? Many functions don't wake up the body so you need to call ```BodyInterface::ActivateBody```. For ```SetAngularVelocity``` there is ```BodyInterface::SetAngularVelocity``` which also wakes up the body.

**Reply by wending10** [reply]
While it is not efficient, i set all my bodies that are dynamic not be able to sleep, the hinges, which is static, are allowed to sleep.
